prov/gni/src/gnix_mr.c

```diff
@@ -574,8 +644,8 @@ static int __mr_cache_register(
 	gni_return_t grc = GNI_RC_SUCCESS;
 
 	/* if we shouldn't introduce any new elements, return -FI_ENOSPC */
-	if (cache->attr.hard_reg_limit > 0 &&
-			atomic_get(&cache->inuse_elements) >= cache->attr.hard_reg_limit)
+	if (unlikely(cache->attr.hard_reg_limit > 0 &&
+			atomic_get(&cache->inuse_elements) >= cache->attr.hard_reg_limit))
 		return FI_ENOSPC;
```

Comment:
Fair enough, I can put this check after we've checked the inuse entries.
-- Jim




On 6/2/15, 11:18 AM, "ztiffany" <notifications@github.com> wrote:

>In 
>prov/gni/src/gnix_mr.c
><https://github.com/ofi-cray/libfabric-cray/pull/202#discussion_r31540725>
>:
>> @@ -574,8 +644,8 @@ static int __mr_cache_register(
>>  	gni_return_t grc = GNI_RC_SUCCESS;
>>  
>>  	/* if we shouldn't introduce any new elements, return -FI_ENOSPC */
>> -	if (cache->attr.hard_reg_limit > 0 &&
>> -			atomic_get(&cache->inuse_elements) >= cache->attr.hard_reg_limit)
>> +	if (unlikely(cache->attr.hard_reg_limit > 0 &&
>> +			atomic_get(&cache->inuse_elements) >= cache->attr.hard_reg_limit))
>>  		return FI_ENOSPC;
>This check can prevent use of an existing MR cache entry that satisfies
>the call when the cache is full.
>—
>Reply to this email directly or
>view it on GitHub 
><https://github.com/ofi-cray/libfabric-cray/pull/202/files#r31540725>.
>
>